consensus/state.go

```diff
@@ -529,12 +529,11 @@ func (cs *State) updateToState(state sm.State) {
 
 	// Reset fields based on state.
 	validators := state.Validators
-	lastPrecommits := (*types.VoteSet)(nil)
 	if cs.CommitRound > -1 && cs.Votes != nil {
 		if !cs.Votes.Precommits(cs.CommitRound).HasTwoThirdsMajority() {
 			panic("updateToState(state) called but last Precommit round didn't have +2/3")
 		}
-		lastPrecommits = cs.Votes.Precommits(cs.CommitRound)
+		cs.LastCommit = cs.Votes.Precommits(cs.CommitRound)
```

Comment:
Could you please explain how this fixes the issue at hand?

Reply:
I did describe the explanation in message commit (sorry to not update the PR description):

```
In State.updateToState, last precommit is computed  only when round
greater than -1 and has votes. But "LastCommit" is always updated
regardless of the condition. If there's no last precommit, "LastCommit"
is set to (*types.VoteSet)(nil). That's why "LastCommit" can be -1 for
heights > 1.

To fix it, only update State.RoundState.LastCommit when there is last
precommit.
```

Reply:
Ok, great

Reply:
> If there's no last precommit, "LastCommit" is set to (*types.VoteSet)(nil). That's why "LastCommit" can be -1 for
heights > 1.

How can there be no last precommit for heights > 1 ?

Reply:
I'm also worried this might mess with patched versions of tendermint that are pruning old blocks cc @hendrikhofstadt

Reply:
> > If there's no last precommit, "LastCommit" is set to (*types.VoteSet)(nil). That's why "LastCommit" can be -1 for
> > heights > 1.
> 
> How can there be no last precommit for heights > 1 ?

We only require +2/3 precommits for a particular block at a round,  so there can be a state that we don't get any votes and `cs.Votes` is `nil`. The if block in the code:

```
if cs.CommitRound > -1 && cs.Votes != nil {
```
indicates that (Maybe you can elaborate more because I see you're the author of those lines).

Reply:
@melekes @ebuchman @hendrikhofstadt any updates on this?

Reply:
I think that's just for the case when we restart the process since in NewConsensusState we call updateToState first and then reconstructLastCommit. I'll need some more time to review though sorry.

Reply:
@ebuchman No rush at all!

Isn't it also happen if we change from fast sync to consensus?

Reply:
Yep that too! It's effectively like a restart (ie. we have some blockchain already but we're only just starting the ConsensusState)

Reply:
> Yep that too! It's effectively like a restart (ie. we have some blockchain already but we're only just starting the ConsensusState)

So it does explains why we have height > 1 but without votes.  When we switch from fast sync to consensus, we call `reconstructLastCommit`, which will set `cs.LastCommit`.

But then we call updateToState, which does not reset state fields, but trigger new round step:

https://github.com/tendermint/tendermint/blob/6a3f75031ccec1acdfc9a2a0ef1f13cc0024e02f/consensus/state.go#L519

So next time we call `updateToState`, `cs.Votes` is still nil, so the if condition does not occur, but we set `cs.LastCommit` (we did set in `reconstructLastCommit`) to nil.

Anyway, I think if we know that the check `if cs.CommitRound > -1 && cs.Votes != nil {` should only happen for restarting, then it's safe to only set `cs.LastCommit` inside the if body.

Reply:
The test log also indicates that it stuck when testing fast sync:

```
Testing fastsync on node 1
local_testnet_1
starting tendermint peer ID=1
741ffbfb263e52d2f6b20f034922c8288604ee11c3263d0cc8061efc7141b34a
starting test client container with CMD=test/p2p/fast_sync/check_peer.sh 1
Other peer is on height 23 with state "0800000000000000"
Waiting for peer 1 to catch up
... 21
... 21
... 21
... 21
... 21
... 21
... 21
... 21
... 21
... 21
... 21
... 21
...
```

Reply:
@ebuchman @melekes Any updates on this?